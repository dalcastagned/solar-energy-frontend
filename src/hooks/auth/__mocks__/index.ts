export const userMock = {
  user: 'teste@teste.com',
  roles: [],
  token:
    'eyJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJBVVJPUkEiLCJhdWQiOiJBVVJPUkEiLCJleHAiOjMwMjc2NTMzNTAsInJvbGVzIjpbIlBBQ1NOR19ESUNPTV9CVUxLREFUQSIsIlBBQ1NOR19ESUNPTV9XQURPIiwiUEFDU05HX0RJQ09NX1FJRE8iLCJQQUNTTkdfQ09ORklHVVJBVElPTiIsIlBBQ1NOR19OT0RFX0NPTlRST0wiLCJQQUNTTkdfU0VSSUVTX01JR1JBVElPTiIsIlBBQ1NOR19TRVJJRVNfUkVNT1ZBTCJdLCJwcmVmZXRjaHJvbGVzIjpbXSwidXNlciI6InVzZXJ0ZXN0ZSJ9.8a_ZdpxfM1_rU-ekCaZnefGFsJVq2_LetC5yBD_80SE',
};

export const loginRequestMock = {
  email: 'usuario@usuario.com',
  password: 'password',
};

