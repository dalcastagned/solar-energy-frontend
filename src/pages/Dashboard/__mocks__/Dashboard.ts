export const counts = {
  activePlants: 0,
  inactivePlants: 0,
};

export const monthsGenerations = [
  {
    month: 'janeiro',
    generation: 150,
  },
];
