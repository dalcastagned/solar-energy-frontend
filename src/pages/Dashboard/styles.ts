import { Alert, Box, Paper } from '@mui/material';

import { styled } from '@mui/material/styles';

export const Wrapper = styled(Box)(() => ({
  display: 'flex',
  width: '100%',
  alignItems: 'flex-end',
  flexDirection: 'column',
  gap: '3rem',
}));

export const ContainerCards = styled(Box)(() => ({
  display: 'flex',
  width: '100%',
  justifyContent: 'center',
  margin: '0 auto',
  flexWrap: 'wrap',
  gap: '2rem',
}));

export const Card = styled(Paper)(() => ({
  maxWidth: '25.8rem',
  height: '13.4rem',
  width: '100%',
  textAlign: 'center',
  border: ' 0.1rem solid #DFE0EB',
  background: '#fff',
  borderRadius: '0.8rem',
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
  justifyContent: 'center',
  cursor: 'pointer',

  '@media screen and (max-width: 580px)': {
    maxWidth: '16rem',
    height: '10rem',
  },

  '&:hover': {
    border: ' 0.1rem solid #3751FF',

    '& h2': {
      color: '#3751FF',
    },

    '& p': {
      color: '#3751FF',
    },
  },

  '& h2': {
    fontSize: '1.9rem',
    fontWeight: '700',
    color: '#9FA2B4',

    '@media screen and (max-width: 580px)': {
      fontSize: '1.6rem',
    },
  },

  '& p': {
    fontSize: '4rem',
    fontWeight: '700',
    color: '#252733',
    paddingTop: '1.2rem',

    '@media screen and (max-width: 580px)': {
      fontSize: '2.2rem',
    },
  },

  '& span': {
    paddingLeft: '0.5rem',
  },
}));

export const ContainerChart = styled(Box)(() => ({
  width: '100%',
  maxWidth: '110rem',
  maxHeight: '64rem',
  border: '1px solid #E5E5E5',
  borderRadius: '0.4rem',
  padding: '3rem',
  margin: '2rem auto 0 auto',

  '& h1': {
    paddingBottom: '3rem',
    fontWeight: '500',
    fontSize: '2rem',
    color: 'rgba(0, 0, 0, 0.87)',
  },
}));

export const Loading = styled(Box)(() => ({
  display: 'flex',
  width: '100%',
  height: 'calc(100vh - 10.4rem)',
  alignItems: 'center',
  justifyContent: 'center',
  flexDirection: 'column',
}));

export const AlertStyled = styled(Alert)(() => ({
  fontSize: '1.6rem',
  width: '100%',
}));
