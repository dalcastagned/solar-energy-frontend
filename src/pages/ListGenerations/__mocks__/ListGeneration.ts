export const generationsMock = {
  hasNextPag: false,
  hasPreviousPage: false,
  isFirstPage: true,
  isLastPage: true,
  pageCount: 1,
  pageNumber: 1,
  pageSize: 1,
  totalItemCount: 1,
  generations: [
    {
      id: 1,
      date: '2022-06-05T17:53:08.189Z',
      generatePower: 150,
    },
  ],
};

export const generationMock = {
  id: 1,
  date: '2022-06-05T17:53:08.189Z',
  generatePower: 150,
};

export const generationsWithPageMock = {
  hasNextPag: false,
  hasPreviousPage: false,
  isFirstPage: true,
  isLastPage: true,
  pageCount: 2,
  pageNumber: 1,
  pageSize: 1,
  totalItemCount: 1,
  generations: [
    {
      id: 1,
      date: '2022-06-05T17:53:08.189Z',
      generatePower: 150,
    },
  ],
};
