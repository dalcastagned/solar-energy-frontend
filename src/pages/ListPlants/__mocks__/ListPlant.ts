export const plantsMock = {
  hasNextPag: false,
  hasPreviousPage: false,
  isFirstPage: true,
  isLastPage: true,
  pageCount: 1,
  pageNumber: 1,
  pageSize: 1,
  totalItemCount: 1,
  plants: [
    {
      id: 1,
      nickname: 'nickname',
      place: 'place',
      brand: 'brand',
      model: 'model',
      active: true,
    },
  ],
};

export const plantMock = {
  id: 1,
  nickname: 'Apelido da Planta',
  place: 'Local da Planta',
  brand: 'Marca da Planta',
  model: 'Modelo da Planta',
  active: true,
};

export const plantsWithPageMock = {
  hasNextPag: false,
  hasPreviousPage: false,
  isFirstPage: true,
  isLastPage: true,
  pageCount: 2,
  pageNumber: 1,
  pageSize: 1,
  totalItemCount: 1,
  plants: [
    {
      id: 1,
      nickname: 'nickname2',
      place: 'place2',
      brand: 'brand2',
      model: 'model2',
      active: false,
    },
  ],
};
