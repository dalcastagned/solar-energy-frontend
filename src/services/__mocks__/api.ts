import axios from 'axios';

export type Error = {
  status?: number;
  data?: Record<string, unknown>;
  message?: string;
};

export type HttpResponse<DataType> = {
  data?: DataType;
  error?: Error;
};

const api = jest.genMockFromModule('axios') as jest.MockedFunction<
  typeof axios
>;

api.create = jest.fn(() => api);

export default api;
