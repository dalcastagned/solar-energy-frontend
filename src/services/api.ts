import axios, { AxiosError, AxiosResponse } from 'axios';

// import { cleanStorage } from '../hooks/auth';

export type Error = {
  status?: number;
  data?: Record<string, unknown>;
  message?: string;
};

export type HttpResponse<DataType> = {
  data?: DataType;
  error?: Error;
};

export type ErrorResponse = {
  response: Error;
  code: string;
};

const responseSuccessHandler = (response: AxiosResponse): AxiosResponse =>
  response;

const forbidden = [401, 403];

const responseErrorHandler = (error: AxiosError): Promise<AxiosError> => {
  const isUnauthorized =
    error.response && forbidden.includes(error.response.status);
  const routers = ['/', '/reset-password'].includes(window.location.pathname);
  if (!routers && isUnauthorized) {
    // cleanStorage();
    window.location.href = '/';
  }

  return Promise.reject(error);
};

const api = axios.create({ baseURL: 'http://localhost:7140/api' });

api.interceptors.response.use(
  response => responseSuccessHandler(response),
  error => responseErrorHandler(error),
);

export default api;
