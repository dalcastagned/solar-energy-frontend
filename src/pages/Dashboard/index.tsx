import { useCallback, useEffect, useState } from 'react';

import { AlertProps, CircularProgress, Typography } from '@mui/material';
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
} from 'chart.js';
import { GenerationPower } from 'models/generations';
import { PlantCounts } from 'models/plants';
import { Line } from 'react-chartjs-2';
import api from 'services/api';
import { handleErrorResponse } from 'utils/error-response';

import * as S from './styles';

type ShowAlertProps = AlertProps & {
  message: string;
};

const Dashboard = (): JSX.Element => {
  const [counts, setCounts] = useState<PlantCounts>();
  const [monthsGenerations, setMonthsGenerations] = useState<GenerationPower[]>(
    [],
  );
  const [alertProps, setAlertProps] = useState<ShowAlertProps>({
    severity: 'success',
    message: '',
  });

  ChartJS.register(
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend,
  );

  const options = {
    responsive: true,
    scales: {
      x: {
        grid: {
          display: false,
        },
      },
    },
    plugins: {
      legend: {
        display: false,
      },
      title: {
        display: false,
      },
    },
  };

  const data = {
    labels: monthsGenerations.map(
      (generation: GenerationPower) => generation.month,
    ),
    datasets: [
      {
        label: 'Geração',
        data: monthsGenerations.map(
          (generation: GenerationPower) => generation.generation,
        ),
        borderColor: '#2196F3',
        backgroundColor: '#2196F3',
      },
    ],
  };

  const getCounts = useCallback(async () => {
    setAlertProps({
      severity: 'info',
      message: '',
    });
    const response = await api.get<PlantCounts>('/plant/counts');
    setCounts(response.data);
  }, []);

  const getGenerations = useCallback(async () => {
    setAlertProps({
      severity: 'info',
      message: '',
    });
    const response = await api.get<GenerationPower[]>(
      '/plant/generations-last-12-months',
    );
    setMonthsGenerations(response.data);
  }, []);

  useEffect(() => {
    getCounts().catch(e =>
      setAlertProps({ severity: 'error', message: handleErrorResponse(e) }),
    );
    getGenerations().catch(e =>
      setAlertProps({ severity: 'error', message: handleErrorResponse(e) }),
    );
  }, [getCounts, getGenerations]);

  return (
    <>
      {!monthsGenerations.length && !counts ? (
        <S.Loading>
          <CircularProgress />
        </S.Loading>
      ) : null}
      {alertProps.message.length ? (
        <S.AlertStyled severity={alertProps.severity}>
          {alertProps.message}
        </S.AlertStyled>
      ) : null}
      {monthsGenerations.length && counts && !alertProps.message.length ? (
        <S.Wrapper>
          <S.ContainerCards>
            <S.Card elevation={2}>
              <Typography variant="h2">Total de Unidades</Typography>
              <p>{counts && counts?.activePlants + counts?.inactivePlants}</p>
            </S.Card>
            <S.Card elevation={2}>
              <Typography variant="h2">Unidades Ativas</Typography>
              <p>{counts?.activePlants}</p>
            </S.Card>
            <S.Card elevation={2}>
              <Typography variant="h2">Unidades Inativas</Typography>
              <p>{counts?.inactivePlants}</p>
            </S.Card>
            <S.Card elevation={2}>
              <Typography variant="h2">Média de Energia</Typography>
              <p>
                {(
                  monthsGenerations.reduce(
                    (acc, curr) => acc + curr.generation,
                    0,
                  ) / 12
                ).toFixed(2)}
                <span>kW</span>
              </p>
            </S.Card>
          </S.ContainerCards>
          <S.ContainerChart>
            <Typography variant="h1">
              Total de energia gerada por mês
            </Typography>
            <Line options={options} data={data} />
          </S.ContainerChart>
        </S.Wrapper>
      ) : null}
    </>
  );
};

export default Dashboard;
