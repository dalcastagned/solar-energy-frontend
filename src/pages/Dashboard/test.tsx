import { screen, waitFor } from '@testing-library/react';
import api from 'services/api';

import Dashboard from '.';
import { renderWithRoute } from '../../utils/tests/helpers';
import { counts, monthsGenerations } from './__mocks__/Dashboard';

describe('<Dashboard />', () => {
  it('should render the main', async () => {
    jest.spyOn(api, 'get').mockImplementation(url => {
      if (url === '/plant/counts') {
        return Promise.resolve({ data: counts });
      }
      return Promise.resolve({ data: monthsGenerations });
    });
    const { container } = renderWithRoute(<Dashboard />);
    await waitFor(() => {
      expect(screen.getByText('Total de Unidades')).toBeInTheDocument();
    });
    expect(container.firstChild).toMatchSnapshot();
  });
  it('should render alert', async () => {
    jest.spyOn(api, 'get').mockRejectedValue({ status: 500 });

    renderWithRoute(<Dashboard />);

    await waitFor(() => {
      expect(
        screen.getByText(
          'O servidor parece estar indisponível. Verifique sua conexão.',
        ),
      ).toBeInTheDocument();
    });
  });
});
