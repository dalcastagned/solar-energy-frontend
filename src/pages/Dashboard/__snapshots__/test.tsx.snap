// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> should render the main 1`] = `
<div
  class="MuiBox-root css-p4ro6g"
>
  <div
    class="MuiBox-root css-1oy7jxl"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-1mhtkfx-MuiPaper-root"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2 css-1sra7t5-MuiTypography-root"
      >
        Total de Unidades
      </h2>
      <p>
        0
      </p>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-1mhtkfx-MuiPaper-root"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2 css-1sra7t5-MuiTypography-root"
      >
        Unidades Ativas
      </h2>
      <p>
        0
      </p>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-1mhtkfx-MuiPaper-root"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2 css-1sra7t5-MuiTypography-root"
      >
        Unidades Inativas
      </h2>
      <p>
        0
      </p>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-1mhtkfx-MuiPaper-root"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2 css-1sra7t5-MuiTypography-root"
      >
        Média de Energia
      </h2>
      <p>
        12.50
        <span>
          kW
        </span>
      </p>
    </div>
  </div>
  <div
    class="MuiBox-root css-14wktby"
  >
    <h1
      class="MuiTypography-root MuiTypography-h1 css-o2w69a-MuiTypography-root"
    >
      Total de energia gerada por mês
    </h1>
    <canvas
      height="150"
      role="img"
      width="300"
    />
  </div>
</div>
`;
